Control/Concurrent/STM/TSem.hs

```diff
@@ -72,11 +77,29 @@ waitTSem (TSem t) = do
   writeTVar t $! (i-1)
 
 
--- TODO: consider using an 'Integer' value for the internal counter
--- value which can never overflow; the computational overhead
--- should hopefully be neglectable in the context of the STM
--- transaction overhead.
+-- | Multi-wait on 'TSem'.
+--
+-- Acquire multiple units at once from semaphore; blocks (via 'retry')
+-- when the number of available units is less than the requested
+-- amount.
+--
+-- > waitTSem == waitTSemN 1
 --
+-- __NOTE__: This operation is subtly different from repeatedly
+-- invoking 'waitTSem'. Note that 'waitTSem 0' can block.
+--
+-- @since TBD
+waitTSemN :: Word.Word -> TSem -> STM ()
+waitTSemN 0 (TSem t) = do
+  i <- readTVar t
+  when (i < 0) retry
+waitTSemN 1 s = waitTSem s
+waitTSemN n (TSem t) = do
+  i <- readTVar t
+  let n' = toInteger n
+  when (i < n') retry
```

Comment:
This offers no fairness: unlike `Control.Concurrent.QSemN`, a single `waitTSemN` can be indefinitely preempted by other `waitTSem` calls, and in fact it probably will be.  I do wonder how useful this is. At the very least we should have a warning in the docs.